File: captainhcg/go/199-binary-tree-right-side-view.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var ret []int
func rightSideView(root *TreeNode) []int {
    ret = make([]int, 0)
    if root == nil{
        return ret
    }
    helper(root, 0)
    return ret
}

func helper(nd *TreeNode, level int){
    if nd == nil{
        return
    }
    if len(ret) <= level{
        ret = append(ret, 0)
    }
    ret[level] = nd.Val
    helper(nd.Left, level+1)
    helper(nd.Right, level+1)
}

File: bwu/binary_tree/199-binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        if not root:
            return res
        level = [root]
        while level:
            n = len(level)
            for i in xrange(n):
                node = level.pop(0)
                if node.left:
                    level.append(node.left)
                if node.right:
                    level.append(node.right)
                if i == n - 1:
                    res.append(node.val)
                    
        return res
